### PR TITLE
Mavlink V1 support for routing.

### PR DESCRIPTION
**Pull request by gbin**
It is a little ugly but works.

### CODE REVIEW

File: src/connection/routing.rs [hunk 1]
```diff
@@ -4,42 +4,111 @@
...
         };
         Ok(len)
     }
 
-    fn raw_read(&self) -> io::Result<MAVLinkV2MessageRaw> {
+    fn raw_read(&self) -> io::Result<MAVLinkMessageRaw> {
         let mut guard = self.reader.lock().unwrap();
         let state = &mut *guard;
         loop {
+            println!("Raw read UDP");
```
Comment (massi31415) [inline src/connection/routing.rs:111]: only for debug no?

Comment (massi31415) [reply src/connection/routing.rs:111]: otherwise it will spam the user terminal

File: src/connection/routing.rs [hunk 2]
```diff
@@ -4,42 +4,111 @@
...
-impl<M: Message> RawMavV2Connection<M> for UdpConnection {
-    fn raw_write(&self, raw_msg: &mut MAVLinkV2MessageRaw) -> io::Result<usize> {
+impl<M: Message> RawConnection<M> for UdpConnection {
+    fn raw_write(&self, msg: &mut dyn CommonMessageRaw) -> io::Result<usize> {
         let mut guard = self.writer.lock().unwrap();
         let state = &mut *guard;
-        #[cfg(feature = "routing")]
-        raw_msg.patch_sequence::<M>(state.sequence);
+        // #[cfg(feature = "routing")]
+        // raw_msg.patch_sequence::<M>(state.sequence);
```
Comment (massi31415) [inline src/connection/routing.rs:96]: Delete?

File: src/connection/routing.rs [hunk 3]
```diff
@@ -53,40 +122,71 @@ impl<M: Message> RawMavV2Connection<M> for UdpConnection {
...
+                return Ok(MAVLinkMessageRaw::V1(msg));
+            } else {
+                if state.recv_buf.slice()[0] != crate::MAV_STX_V2 {
+                    println!("Invalid MAVLink magic"); // TODO(gbin): log
+                    continue;
+                }
+                let Ok(msg) = read_v2_raw_message(&mut state.recv_buf) else {
+                 println!("Error reading message from UDP");// TODO(gbin): log
+                 continue;
+             };
```
Comment (massi31415) [inline src/connection/routing.rs:139]: indent?

File: src/connection/routing.rs [hunk 4]
```diff
@@ -53,40 +122,71 @@ impl<M: Message> RawMavV2Connection<M> for UdpConnection {
...
+                if !msg.has_valid_crc::<M>() {
+                    println!("Invalid CRC"); // TODO(gbin): log
+                    continue;
+                }
```
Comment (massi31415) [inline src/connection/routing.rs:143]: This can be check for V1 too

File: src/connection/routing.rs [hunk 5]
```diff
@@ -53,40 +122,71 @@ impl<M: Message> RawMavV2Connection<M> for UdpConnection {
...
-        let mut sequence = self.sequence.lock().unwrap();
-        raw_msg.patch_sequence::<M>(*sequence);
-        *sequence = sequence.wrapping_add(1);
-
-        let l = raw_msg.len();
-        match (&mut *port).write_all(&raw_msg.0[..l]) {
-            Ok(_) => Ok(l),
+        //let mut sequence = self.sequence.lock().unwrap();
+        //raw_msg.patch_sequence::<M>(*sequence);
+        //*sequence = sequence.wrapping_add(1);
```
Comment (massi31415) [inline src/connection/routing.rs:159]: Left over to delete

File: src/connection/routing.rs [hunk 6]
```diff
@@ -53,40 +122,71 @@ impl<M: Message> RawMavV2Connection<M> for UdpConnection {
...
 }
 
-impl<M: Message> RawMavV2Connection<M> for SerialConnection {
-    fn raw_write(&self, raw_msg: &mut MAVLinkV2MessageRaw) -> io::Result<usize> {
+impl<M: Message> RawConnection<M> for SerialConnection {
+    fn raw_write(&self, msg: &mut dyn CommonMessageRaw) -> io::Result<usize> {
         let mut port = self.port.lock().unwrap();
-        let mut sequence = self.sequence.lock().unwrap();
-        raw_msg.patch_sequence::<M>(*sequence);
-        *sequence = sequence.wrapping_add(1);
```
Comment (massi31415) [inline src/connection/routing.rs:71]: We don't want to patch the seq # here?

File: src/connection/routing.rs [hunk 7]
```diff
@@ -53,40 +122,71 @@ impl<M: Message> RawMavV2Connection<M> for UdpConnection {
...
 
         loop {
-            let Ok(raw_message) = read_v2_raw_message(&mut *port) else {
+            let Ok(msg) = read_raw_message::<SystemPort,M>(&mut *port) else {
                 continue;
             };
-            if raw_message.has_valid_crc::<M>() {
-                return Ok(raw_message);
-            }
+            match msg {
```
Comment (massi31415) [inline src/connection/routing.rs:173]: I would implement this in the `CommonMessageRaw` trait for the `MAVLinkMessageRaw` type? i.e. add has_valid_crc to the trait

File: src/connection/udp.rs [hunk 8]
```diff
@@ -47,7 +47,7 @@ pub fn udpout<T: ToSocketAddrs>(address: T) -> io::Result<UdpConnection> {
         .next()
         .expect("Invalid address");
     let socket = UdpSocket::bind("0.0.0.0:0")?;
-    UdpConnection::new(socket, false, Some(addr))
+    UdpConnection::new(socket, false, Some(addr), addr.to_string().as_str())
```
Comment (massi31415) [inline src/connection/udp.rs:50]: why adding `addr` as str to the args?

Comment (massi31415) [reply src/connection/udp.rs:50]: Ah I see you redefined the constructor bellow.

Comment (massi31415) [reply src/connection/udp.rs:50]: It is confusing bit here. Maybe you can put `addr.to_string().as_str()` in a const and give it appropriate naming to explain the functionality i.e. it is an id or something

File: src/lib.rs [hunk 9]
```diff
@@ -211,83 +212,107 @@ pub fn read_versioned_msg<M: Message, R: Read>(
 }
 
 #[derive(Debug, Copy, Clone, PartialEq, Eq)]
-// Follow protocol definition: `<https://mavlink.io/en/guide/serialization.html#v1_packet_format>`
-pub struct MAVLinkV1MessageRaw([u8; 1 + Self::HEADER_SIZE + 255 + 2]);
+pub struct MAVLinkAnyMessageRaw<const S: usize>([u8; S]);
```
Comment (massi31415) [inline src/lib.rs:215]: I would add the following fields for conveniency: `HEADER_SIZE`, `MAX_SIZE`

File: src/lib.rs [hunk 10]
```diff
@@ -211,83 +212,107 @@ pub fn read_versioned_msg<M: Message, R: Read>(
...
+    fn len(&self) -> usize;
+    fn payload_length(&self) -> usize;
+    fn payload(&self) -> &[u8];
+    fn full(&self) -> &[u8];
+}
+
+pub type MAVLinkV1MessageRaw = MAVLinkAnyMessageRaw<MAX_SIZE_V1>;
+
+const HEADER_SIZE_V1: usize = 5;
+pub const MAX_SIZE_V1: usize = 1 + HEADER_SIZE_V1 + 255 + 2;
```
Comment (massi31415) [inline src/lib.rs:232]: Let add those to the `MAVLinkV1MessageRaw ` struct and expose them as pub?

Comment (massi31415) [reply src/lib.rs:232]: this means that we need to implement the struct first then the trait

File: src/lib.rs [hunk 11]
```diff
@@ -211,83 +212,107 @@ pub fn read_versioned_msg<M: Message, R: Read>(
...
 
     #[inline]
     fn mut_payload_and_checksum(&mut self) -> &mut [u8] {
-        let payload_length: usize = self.payload_length().into();
-        &mut self.0[(1 + Self::HEADER_SIZE)..(1 + Self::HEADER_SIZE + payload_length + 2)]
+        let payload_length = self.payload_length();
+        &mut self.0[(1 + HEADER_SIZE_V1)..(1 + HEADER_SIZE_V1 + payload_length + 2)]
     }
 
     pub fn calculate_crc<M: Message>(&self) -> u16 {
```
Comment (massi31415) [inline src/lib.rs:312]: I am wondering how much is the overhead here. Specially at the FC side where compute is limited !

File: src/lib.rs [hunk 12]
```diff
@@ -377,9 +398,54 @@ const SIGNATURE_SIZE_V2: usize = 13;
 
 pub const MAX_SIZE_V2: usize = 1 + HEADER_SIZE_V2 + 255 + 2 + SIGNATURE_SIZE_V2;
 
-#[derive(Debug, Copy, Clone, PartialEq, Eq)]
+pub type MAVLinkV2MessageRaw = MAVLinkAnyMessageRaw<MAX_SIZE_V2>;
```
Comment (massi31415) [inline src/lib.rs:401]: I would add the HEADER_SIZE, MAX_SIZE as fields

**Review by massi31415 — CHANGES_REQUESTED** [review]
left few comments !

File: src/connection/routing.rs (hunk 1 above)
Comment (gbin) [reply src/connection/routing.rs:111]: fixed later

File: src/connection/routing.rs (hunk 2 above)
Comment (gbin) [reply src/connection/routing.rs:96]: removed later yes, this is breakig mavlink contract for signature

File: src/connection/routing.rs (hunk 3 above)
Comment (gbin) [reply src/connection/routing.rs:139]: got fixed with logging

File: src/connection/routing.rs (hunk 5 above)
Comment (gbin) [reply src/connection/routing.rs:159]: deleted later

File: src/lib.rs (hunk 12 above)
Comment (gbin) [reply src/lib.rs:401]: I just moved them out so it can be reused.

File: src/lib.rs (hunk 11 above)
Comment (gbin) [reply src/lib.rs:312]: &mut self.0[(1 + HEADER_SIZE_V1)..(1 + HEADER_SIZE_V1 + payload_length + 2)]  this is almost 0 overhead this is just a view/slice on the message

File: src/connection/udp.rs (hunk 8 above)
Comment (gbin) [reply src/connection/udp.rs:50]: Good point, this will be refactored, I hate those string copies everywhere.

**Pull request merged**